test/security/CsrCredentialsProviderSpec.scala

```diff
@@ -0,0 +1,174 @@
+import uk.gov.hmrc.play.http.HeaderCarrier
+
+import scala.concurrent.Future
+
+class CsrCredentialsProviderSpec extends BaseSpec with ScalaFutures {
+
+  "authenticate" should {
+    "return InvalidRole if the role is invalid" in new TestFixture {
+      val csrCredentialsProvider = new CsrCredentialsProvider {
+        val http = CSRHttp
+        override def signIn(email: String, password: String)(implicit hc: HeaderCarrier): Future[UserResponse] = {
+          Future.failed(new InvalidRoleException)
+        }
+      }
+
+      val result = csrCredentialsProvider.authenticate(Credentials(Id, Password))
+      result.futureValue mustBe Left(InvalidRole)
+    }
+
+    "return InvalidCredentials when the credentials are invalid and we cannot record login failure" in new TestFixture {
+      val csrCredentialsProvider = new CsrCredentialsProvider {
+        val http = CSRHttp
+        override def signIn(email: String, password: String)(implicit hc: HeaderCarrier): Future[UserResponse] = {
+          Future.failed(new InvalidCredentialsException)
+        }
+        override def failedLogin(email: String)(implicit hc: HeaderCarrier): Future[UserResponse] = {
+          Future.failed(new InvalidCredentialsException())
+        }
+      }
+      val result = csrCredentialsProvider.authenticate(Credentials(Id, Password))
+      result.futureValue mustBe Left(InvalidCredentials)
+    }
+
+    "return InvalidCredentials when the credentials are invalid" in new TestFixture {
+      val csrCredentialsProvider = new CsrCredentialsProvider {
+        val http = CSRHttp
+        override def signIn(email: String, password: String)(implicit hc: HeaderCarrier): Future[UserResponse] = {
+          Future.failed(new InvalidCredentialsException)
+        }
+        override def failedLogin(email: String)(implicit hc: HeaderCarrier): Future[UserResponse] = {
+          val userResponse = UserResponse(FirstName, LastName, Some(PreferredName), true, UserId,
+            Email, "", Role, ServiceName)
+          Future.successful(userResponse)
+        }
+      }
+      val result = csrCredentialsProvider.authenticate(Credentials(Id, Password))
+      result.futureValue mustBe Left(InvalidCredentials)
+    }
+
+    "return user when credentials are valid" in new TestFixture {
+      val csrCredentialsProvider = new CsrCredentialsProvider {
+        val http = CSRHttp
+        override def signIn(email: String, password: String)(implicit hc: HeaderCarrier): Future[UserResponse] = {
+          val userResponse = UserResponse(FirstName, LastName, Some(PreferredName), true, UserId,
+            Email, "", Role, ServiceName)
+          Future.successful(userResponse)
+        }
+
+      }
+      val result = csrCredentialsProvider.authenticate(Credentials(Id, Password))
+      result.futureValue mustBe Right(CachedUser(UserId, FirstName, LastName, Some(PreferredName), Email, true, ""))
+    }
+  }
+  "recordFailedAttempt" should {
```

Comment:
Not sure your opinion, but I would say that recordFailedAttempt should be a private method, as it is not intended to be called directly. Therefore the authenticate method should be tested for lockout scenarios, not the recordFailedAttempt. It would decouple the test from the implementation and just bind it to the interface that is used by clients of CsrCredentialsProvider. 
One option to pursue is to create a TestCsrCredentialsProvider which just takes the responses from API calls as parameters. It would reduce duplication in the tests significantly